stl/inc/condition_variable

```diff
@@ -128,6 +131,58 @@ public:
         return true;
     }
 
+#if _HAS_CXX20
+private:
+    struct _Cv_any_notify_all {
+        condition_variable_any* _This;
+
+        _Cv_any_notify_all(condition_variable_any* _This_) : _This{_This_} {}
+
+        _Cv_any_notify_all(const _Cv_any_notify_all&) = delete;
+        _Cv_any_notify_all& operator=(const _Cv_any_notify_all&) = delete;
+
+        void operator()() const noexcept {
+            _This->notify_all();
+        }

```

Comment:
I think you need to take out a lock on the internal mutex before calling `notify_all()` here to avoid a missed wake-up.

eg. consider in the `wait()` implementation below that the call to `.request_stop()` occurs after the check of `_Stoken.stop_requested()` and before the call to `wait(_Lck)`.

Reply:
`notify_all` already takes the internal mutex.

Reply:
Maybe we need to take the internal mutex before creating _Cb below...

Reply:
That would explain why this was added to cv_any and not plain cv.

Although I think there might be a spec defect here: If the registration for cancel is supposed to be part of the atomic operation that includes unlocking `_Lck` and going to sleep, the spec doesn't say that right now.

Reply:
Yes, this is broken in the spec. I asked SG1 to comment:

```
{
    condition_variable_any cv;
    mutex m;
    jthread j{[&](stop_token token) {
        unique_lock lck{m};
        cv.wait(token, [] { return false; });
    }};
} // destroy j
```

This program can deadlock under the current spec; there is nothing forbidding the following execution:

T1: launches T2  
T2: takes `m`
T2: [thread.condvarany.intwait]/2 "Registers for the duration of this call `*this` to get notified on a stop request on `stoken` during this call", stop is not requested so execution continues
T2: `while (!stoken.stop_requested()) {`  
T2: `if (pred())` (`pred` returns `false`)  
T1: `j.~jthread`  
T1: `request_stop()`, calls `cv.notify_all()`  
T2: `wait(lck)`  
T1: `join()`  
Deadlock.

Reply:
> This program can deadlock under the current spec

Can you elaborate on what causes the deadlock in this example?
I'm struggling to see where the deadlock occurs.
AFAICT, if `request_stop()`, and thus `cv.notify_all()`, has returned then T1 is no longer holding any resources that T2 is waiting on.

> Maybe we need to take the internal mutex before creating _Cb below...

You don't want to do that. The callback might execute inline during construction of _Cb, which would then call notify_all(), which then also takes out a lock on the internal mutex.

Reply:
>Can you elaborate on what causes the deadlock in this example?

After the execution I enumerated, T2 is waiting on the CV, and T1 is waiting on T2, and nothing will ever notify the CV.

>if request_stop(), and thus cv.notify_all(), has returned then T1 is no longer holding any resources that T2 is waiting on.

Correct, T1 is not holding any such resources. T2 isn't waiting on resources. But T2 is waiting to be notified.

>You don't want to do that. The callback might execute inline during construction of _Cb, which would then call notify_all(), which then also takes out a lock on the internal mutex. 

I see, we need something more complex :/

Reply:
The reference implementation indeed does something a bit tricker to ensure it doesn't miss wakeup notifications.
See https://github.com/josuttis/jthread/blob/master/source/condition_variable_any2.hpp#L216-L238

It first locks the internal mutex and then checks `stoken.stop_requested()`.
This prevents the missed wakeup if `request_stop()` is called between the evaluation of `pred()` and `wait(lck)`.
The registered callback will block until `wait()` releases the internal mutex before calling `.notify_all()`.

Reply:
>The reference implementation indeed does something a bit tricker to ensure it doesn't miss wakeup notifications.

I'm aggressively trying to not look at the reference implementation because licensing scary.

>It first locks the internal mutex and then checks stoken.stop_requested().

Right, that's the exact fix. But nothing in the spec requires that right now.